### PR TITLE
Introduce Role-Based Authentication for Repository Management

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
To address the migration of mirroring configurations from projects to repositories and to simplify access management,
we decided to:
- Hide the meta repository.
- Replace permission-based authentication with a role-based system using `RepositoryRole` (`READ`, `WRITE`, `ADMIN`). The `ADMIN` role has access to the configurations.

This change resolves the issue where users with `WRITE` permission for the meta repository were only allowed to creating mirroring configurations.

With `RepositoryRole`, access becomes more structured and extensible for future enhancements (e.g., introducing custom roles). While permission-based authentication is replaced here, this does not preclude the future coexistence of role-based and permission-based systems.

Modifications:
- Added `RepositoryRole` with hierarchical roles (`READ`, `WRITE`, `ADMIN`).
- Replaced `RequiresPermission` annotations with `RequiresRepositoryRole`.
- Renamed `RequiresRole` to `RequiresProjectRole`.
  - Updated it to accept a single `ProjectRole`, utilizing the hierarchical model to avoid redundant role specifications.
- Removed APIs for managing permissions, replacing them with role management APIs.

Result:
- Role-based access simplifies management and aligns with repository-specific mirroring configurations.
- (Breaking) APIs for managing permissions are removed.

To-do:
- Update the documentation to reflect the new systme, including updated screenshots

Migration plan:
- Deploy [PR](https://github.com/line/centraldogma/pull/1061), which supports deserialization of both legacy and new metadata format.
- Deploy intermediate commit supporting both permission and role APIs, ensuring metadata is stored in the new format.
  - The commit also migrate the legacy format to new format.
- Deploy this commit, which exclusively supports role-based APIs.

### CODE REVIEW

File: server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java [hunk 1]
```diff
@@ -145,17 +144,10 @@ public CompletableFuture<ProjectDto> createProject(CreateProjectRequest request,
...
      *
      * <p>Gets the {@link ProjectMetadata} of the specified {@code projectName}.
-     * If a {@code checkPermissionOnly} parameter is {@code true}, it is returned whether the user has
-     * permission to read the metadata of the specified {@code projectName}.
      */
     @Get("/projects/{projectName}")
-    @RequiresRole(roles = { ProjectRole.OWNER, ProjectRole.MEMBER })
-    public CompletableFuture<ProjectMetadata> getProjectMetadata(
-            @Param String projectName,
-            @Param("checkPermissionOnly") @Default("false") boolean isCheckPermissionOnly) {
```
Comment (minwoox) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java:155]: This parameter was introduced [here](https://github.com/line/centraldogma/pull/133/files#diff-da7ebe272e478dcd441cb91f2cd7680672cfabd20b271a9c970b7bb14facc52bR7) but has not been used anywhere. Let me just remove that.

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java [hunk 2]
```diff
@@ -673,128 +706,127 @@ public CompletableFuture<Revision> updateWriteQuota(
...
-        switch (role) {
-            case OWNER:
-                return repositoryMetadata.perRolePermissions().owner();
-            case MEMBER:
-                return repositoryMetadata.perRolePermissions().member();
-            default:
-                return repositoryMetadata.perRolePermissions().guest();
+    @Nullable
+    private static RepositoryRole repositoryRole(Roles roles, @Nullable RepositoryRole repositoryRole,
+                                                 @Nullable ProjectRole projectRole) {
```
Comment (ikhoon) [inline server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java:801]: `projectRole` seems not null always in the current usage.
```suggestion
                                                 ProjectRole projectRole) {
```

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java [hunk 3]
```diff
@@ -673,128 +706,127 @@ public CompletableFuture<Revision> updateWriteQuota(
...
+            memberOrGuestRole = roles.projectRoles().guest();
+        }
+
+        if (repositoryRole == null) {
+            return memberOrGuestRole;
+        }
+        if (memberOrGuestRole == null) {
+            return repositoryRole;
+        }
+        if (memberOrGuestRole == RepositoryRole.WRITE || repositoryRole == RepositoryRole.WRITE) {
```
Comment (ikhoon) [inline server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java:819]: Let's say `a-member-user` is granted to `READ` role and `projects.member` is granted to `WRITE`. Is `WRITE` given for `a-member-user`?
```json
{
  "repos": {
    "my-repo": {
      "name": "my-repo",
      "roles": {
        "users": {
          "a-member-user": "READ"
        }
        "projects": {
          "member": "WRITE",
          ...
       }
      }
    }
  }
}
```

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/Roles.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+     * Returns the {@link RepositoryRole}s of tokens.
+     */
+    @JsonProperty("tokens")
+    public Map<String, RepositoryRole> tokens() {
+        return tokens;
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this).omitNullValues()
```
Comment (ikhoon) [inline server/src/main/java/com/linecorp/centraldogma/server/metadata/Roles.java:77]: nit:
```suggestion
        return MoreObjects.toStringHelper(this)
```

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java (hunk 3 above)
Comment (minwoox) [reply server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java:819]: > Is WRITE given for a-member-user?

That is correct, as similar to how GitHub handles it.
The role with the highest authority will take precedence.

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java [hunk 5]
```diff
@@ -397,38 +411,52 @@ public CompletableFuture<Revision> restoreRepo(Author author, String projectName
...
+        //         "tokens": ...
+        //         "projects": {
+        //           "member": "READ",
+        //           "guest": null
+        //        }
+        //       }
+        //     }
+        //   }
+        // }
+        final JsonPointer path = JsonPointer.compile("/repos" + encodeSegment(repoName) + "/roles/projects");
```
[the remark this answers is not in the document]
Comment (minwoox) [reply server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java:454]: That's a good suggestion. By the way, this PR is in WIP and I forgot to tell you about it.
Would you mind reviewing https://github.com/line/centraldogma/pull/1064, beforehand?

**Comment by minwoox** [comment]
This PR is ready. PTAL. 🙇

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java [hunk 6]
```diff
@@ -900,90 +842,95 @@ public CompletableFuture<Revision> updateWriteQuota(
...
-    private static Collection<Permission> findPerRolePermissions(RepositoryMetadata repositoryMetadata,
-                                                                 ProjectRole role) {
-        switch (role) {
-            case OWNER:
-                return repositoryMetadata.perRolePermissions().owner();
-            case MEMBER:
-                return repositoryMetadata.perRolePermissions().member();
-            default:
-                return repositoryMetadata.perRolePermissions().guest();
+    @Nullable
```
Comment (jrhee17) [inline server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java:903]: Optional) What do you think of adding a `RepositoryRole.NONE`? It doesn't seem intuitive that a method returning an enum is nullable.

```suggestion
```

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java (hunk 3 above)
Comment (jrhee17) [reply server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java:923]: That's a little surprising - I was expecting the repository role to override the project role.
If that's how Github is doing it, then I guess the current approach is fine.

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java [hunk 7]
```diff
@@ -900,90 +842,95 @@ public CompletableFuture<Revision> updateWriteQuota(
...
+    private static RepositoryRole repositoryRole(Roles roles, @Nullable RepositoryRole repositoryRole,
```
Comment (jrhee17) [inline server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java:904]: optional) the logic in this class was not easy to follow at first glance. To me, the following flow makes more sense:

```
    @Nullable
    private static RepositoryRole repositoryRole(Roles roles, @Nullable RepositoryRole repositoryRole,
                                                 ProjectRole projectRole) {
        final RepositoryRole projectRepositoryRole;
        if (projectRole == ProjectRole.MEMBER) {
            projectRepositoryRole = roles.projectRoles().member();
        } else {
            projectRepositoryRole = roles.projectRoles().guest();
        }
        if (repositoryRole == RepositoryRole.ADMIN || projectRepositoryRole == RepositoryRole.ADMIN) {
            return RepositoryRole.ADMIN;
        }
        if (repositoryRole == RepositoryRole.WRITE || projectRepositoryRole == RepositoryRole.WRITE) {
            return RepositoryRole.ADMIN;
        }
        if (repositoryRole == RepositoryRole.READ || projectRepositoryRole == RepositoryRole.READ) {
            return RepositoryRole.READ;
        }
        return null;
    }
```

**Review by jrhee17 — APPROVED** [review]
Apart from `MetadataService#repositoryRole`, I understood that the rest of changes are simple renaming/refactoring.

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java (hunk 6 above)
Comment (minwoox) [reply server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java:903]: One reason I avoided using `RepositoryRole.NONE` is that it shouldn't be used when assigning a role to a member or a token.
If you're not strongly opposed, let me return `null`. 😉

File: server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java (hunk 7 above)
Comment (minwoox) [reply server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java:904]: Good suggestion. Applied. 👍

**Review by trustin — APPROVED** [review]
An exciting change 🎉

**Pull request merged**